ui/components/HeadProvider/index.tsx

```diff
@@ -0,0 +1,188 @@
+import React, { useContext, createContext, ReactNode, useState } from 'react'
+import { globalHistory } from '@reach/router'
+
+interface ScriptTagParams {
+  type: 'script'
+  id: string
+  script: string
+  texttype: string
+}
+
+interface TitleTagParams {
+  type: 'title'
+  id: string
+  text: string
+}
+
+interface MetaTagParams {
+  type: 'meta'
+  id: string
+  name: 'description' | 'keywords' | 'author' | 'viewport' | 'robots' | 'googlebot'
+  content: string
+}
+
+type TagParams = ScriptTagParams | TitleTagParams | MetaTagParams
+
+type NewTagParams =
+  | Omit<ScriptTagParams, 'id'>
+  | Omit<TitleTagParams, 'id'>
+  | Omit<MetaTagParams, 'id'>
+
+export interface Head {
+  tags: React.ReactElement[]
+  addTag: (params: TagParams) => void
+}
+
+const HeadContext = createContext<Head>({
+  tags: [],
+  addTag: () => {},
+})
+
+interface HeadProviderProps {
+  children: ReactNode
+  tags: React.ReactElement[]
+}
+
+export function HeadProvider(props: HeadProviderProps) {
+  let { children, tags } = props
+  globalHistory.listen(() => {
+    if (typeof window !== 'undefined') {
+      const metaElements = Array.from(
+        document.getElementsByTagName('head')[0].getElementsByTagName('meta'),
+      )
+
+      const titleElements = Array.from(
+        document.getElementsByTagName('head')[0].getElementsByTagName('title'),
+      )
+
+      const scriptElements = Array.from(
+        document.getElementsByTagName('head')[0].getElementsByTagName('script'),
+      )
+
+      metaElements.map((itm) => {
+        if (itm.parentNode && itm.dataset.id) itm.parentNode.removeChild(itm)
+      })
+
+      titleElements.map((itm) => {
+        if (itm.parentNode && itm.dataset.id) itm.parentNode.removeChild(itm)
+      })
+
+      scriptElements.map((itm) => {
+        if (itm.parentNode && itm.dataset.id) itm.parentNode.removeChild(itm)
+      })
+    } else {
+    }
+  })
+  return (
+    <HeadContext.Provider
+      value={{
+        tags,
+        addTag: ({ id, ...params }) => {
+          if (params.type === 'title') {
+            if (typeof window !== 'undefined') {
+              const currentElement =
+                (document.querySelector(`[data-id="${id}"]`) as HTMLTitleElement) || undefined
+              if (currentElement) {
+                if (currentElement.innerHTML === params.text) return
+                currentElement.innerHTML = params.text
+                return
+              }
+
+              const elem = document.createElement('title')
+              elem.innerText = params.text
+              elem.dataset.id = id
+              document.getElementsByTagName('head')[0].appendChild(elem)
+            } else {
+              if (tags.find((itm) => itm.props.children === params.text && itm.type === 'title'))
```

Comment:
Is the ID here as well, rather than type and text.

Reply:
This applies to all the tag types, if that's not obvious.

Reply:
No, that's the issue. On server rerenders it's new IDs but same text and other stuff

Reply:
For example if you add ID the query page will have double of everything with separate IDs due to rerender after query returns.

Reply:
As long as React renders the components in the same order on the client (which it will) the IDs should match up.